### PR TITLE
Read MPAS 

### PR DESCRIPTION
**Pull request by philipc2**
### Overview
Provides support for reading in MPAS grid datasets and encoding them in the UGRID conventions, written according to the MPAS specifications outlined in [this](https://mpas-dev.github.io/files/documents/MPAS-MeshSpec.pdf) document. 
* Similar approach to what is already implemented for other grid conventions (`_scrip.py`, `_exodus.py`)
* Supports converting both the Primal and Dual Meshes to the UGRID conventions.
* Assigns all required UGRID variables for a 2D topology, with the addition of `Mesh2_face_x/y` and `Mesh2_edge_nodes` when available 

### Usage
```Python
import xarray as xr 
import uxarray as ux

# path to grid file in the mpas format
mpas_grid_path = "/path/to/mpas/grid.nc"

# load with xarray 
mpas_xr_ds = xr.open_dataset(mpas_grid_path)

# load the primal mesh
ugrid_primal = ux.Grid(mpas_xr_ds)
# alternatively 
# ugrid_primal = ux.Grid(mpas_xr_ds, use_dual=False)

# load the dual mesh 
ugrid_dual = ux.Grid(mpas_xr_ds, use_dual=True)

### CODE REVIEW

**Comment by rljacob** [comment]
Are the contents of an MPAS grid file specified somewhere?

**Comment by philipc2** [comment]
> Are the contents of an MPAS grid file specified somewhere?

https://mpas-dev.github.io/files/documents/MPAS-MeshSpec.pdf

**Comment by philipc2** [comment]
Hi @cdeciampa 

Considering your recent work with visualizing MPAS grids, I'd like to get your input on converting from MPAS to UGRID. Let me know if you have any questions!

**Review by cdeciampa — COMMENTED** [review]
Looks good, the only suggestion I have is to convert lonCell, latCell, lonVertex, and latVertex from radians to degrees. I'm curious, though, why use the dual-mesh instead of the primary mesh?

**Comment by philipc2** [comment]
> Looks good, the only suggestion I have is to convert lonCell, latCell, lonVertex, and latVertex from radians to degrees. I'm curious, though, why use the dual-mesh instead of the primary mesh?

Whoops! I think I mixed up the Dual and Primary meshes when writing the comments. This implementation should be from the _Primary_ mesh, not the Dual one. Good catch!

**Comment by rljacob** [comment]
@xylar check this out.

File: uxarray/_mpas.py
Comment (xylar) [inline uxarray/_mpas.py:123]: Iterating over cells in python is quite slow.  I have always iterated over the `maxEdges` dimension instead in loops like this one:
```suggestion
    maxEdges = verticesOnCell.shape[1]
    for vert_idx in range(maxEdges):
        mask = vert_idx < nEdgesOnCell
        # convert non-padded values to zero-index
        verticesOnCell[mask, vert_idx] -= 1
        # replace remaining padding or zeros with fill_value
        verticesOnCell[np.logical_not(mask), vert_idx] = fill_val
```

File: uxarray/helpers.py
Comment (xylar) [inline uxarray/helpers.py:45]: It looks like the docstring needs to be updated to include MPAS (and maybe remove shape files, which don't in fact seems to be supported yet).

Also, is there any concern that the approach of looking for a specific variable name to detect a given format won't be robust?  `verticesOnCell` seems pretty safe for detecting MPAS but `coord` and `coordx` seem like they could be pretty generic names.  I know from ParaView that it can be frustrating to have files misidentified because the reader's identification of a format is too generic (every NetCDF file gets identified as a CAM atmosphere file for some reason).

File: test/test_mpas.py
Comment (xylar) [inline test/test_mpas.py:16]: These examples are pretty big files.  In my experience, having multi MB files in a repo gets cumbersome.  I would suggest something much coarser.  As examples, we have this very coarse mesh in MPS-Tools:
https://github.com/MPAS-Dev/MPAS-Tools/blob/master/mesh_tools/mesh_conversion_tools/test/mesh.QU.1920km.151026.nc

**Review by xylar — COMMENTED** [review]
This looks really exciting!

I do think having support for the dual mesh will be quite important, as we often want to manipulate and visualize quantities on both meshes.  But the primary mesh is certainly the right starting point.

File: uxarray/_mpas.py
Comment (xylar) [reply uxarray/_mpas.py:123]: I ran `pytest test` locally with these changes and it passes (not sure how rigorous a test that is).

I'm happy to do some other testing but not sure how much can be done with uxarray yet.  The docs don't seem to go much beyond reading in files and getting connectivity info in various ways. (I fully understand that this is a work in progress.)

Comment (cdeciampa) [reply uxarray/_mpas.py:123]: Now that I'm taking a closer look at this (it's been a long week), I'm noticing that within `verticesOnCell`, the shapes that have sides n < 7 are all padded out with the same value, which doesn't actually exist as a vertex index in the mesh file. While I was trying to test this on my own with Polymesh, I ran into downstream issues with numpy indexing for coordinate values (ie; `lonCell[verticesOnCell]`, `latCell[verticesOnCell]`).

I've done this differently within my own code by having each individual cell connect back to each individual "origin" vertex. It's a little clunky, but is marginally faster this way as well (~20-30 ms):

```
%%time
# MPAS indexing starts at 1, not 0 (any original 0 value ~ NaN)
face_nodes = c_mesh_ds.verticesOnCell.values - 1

# Replaces all -1 values with first value in array to force 7-sided shapes
face_nodes[:, 4] = np.where(face_nodes[:, 4] == -1, face_nodes[:, 0], face_nodes[:, 4])
face_nodes[:, 5] = np.where(face_nodes[:, 5] == -1, face_nodes[:, 0], face_nodes[:, 5])
face_nodes[:, 6] = np.where(face_nodes[:, 6] == -1, face_nodes[:, 0], face_nodes[:, 6])

# Ensures face nodes are returned as integers
face_nodes = face_nodes.astype(np.uint32)
```

```
CPU times: user 27.9 ms, sys: 32.2 ms, total: 60.1 ms
Wall time: 58.2 ms
```

Comment (xylar) [reply uxarray/_mpas.py:123]: @cdeciampa, yes, that's another approach that's good for visualization (and I believe also the approach expected by the SCRIP format).  I don't know if that's an appropriate solution for uxarray more broadly.  That presumably needs to be a decision on what the standard uxarray format should be.

File: test/test_mpas.py
Comment (rljacob) [reply test/test_mpas.py:16]: I agree we should avoid large binary files in a git repo.

File: uxarray/helpers.py
Comment (philipc2) [reply uxarray/helpers.py:45]: > It looks like the docstring needs to be updated to include MPAS (and maybe remove shape files, which don't in fact seems to be supported yet).
> 
> Also, is there any concern that the approach of looking for a specific variable name to detect a given format won't be robust? `verticesOnCell` seems pretty safe for detecting MPAS but `coord` and `coordx` seem like they could be pretty generic names. I know from ParaView that it can be frustrating to have files misidentified because the reader's identification of a format is too generic (every NetCDF file gets identified as a CAM atmosphere file for some reason).

Good catch on the docstring. 

I agree that the way that we check what format the grid is in is not the most robust. I'll raise an issue about it today, since having more complete functions to check what format a dataset is in would be better (i.e. `_is_mpas()`, `_is_scrip()`, etc)

**Comment by philipc2** [comment]
> This looks really exciting!
> 
> I do think having support for the dual mesh will be quite important, as we often want to manipulate and visualize quantities on both meshes. But the primary mesh is certainly the right starting point.

@erogluorhan @rajeeja 

How would you suggest handling both meshes within `UXarray`? Since we currently have a singular dataset under our `Grid` object, storing both the Primal and Dual mesh would not be ideal. Should we add a parameter to let the user chose which grid to load in?

File: uxarray/_mpas.py
Comment (xylar) [reply uxarray/_mpas.py:123]: @philipc2, my version accounts for all of that.  The mask array will be true for cells that have a vertex with the given index and will be false for cells that don't have it.  It is vastly more efficient to iterate over 7 to 9 vertices than it is to iterate over potentially tens of millions of cells.  So I stand by my version being much preferable, especially for very large MPAS meshes.  The result should be the same as your loop, as you can check.

**Comment by rljacob** [comment]
If a file has multiple meshes in it, then yes there should be a parameter to let the user specify which one to load.  And if they go back to that mesh file for a second mesh, thats a second grid object. 

MPAS files I assume have easy ways to find one or the other mesh.  Not sure how to do that in general

File: uxarray/helpers.py
Comment (philipc2) [reply uxarray/helpers.py:45]: #238

**Comment by philipc2** [comment]
@xylar 

All changes should be addressed now, with the added support for the Dual mesh.

File: uxarray/_mpas.py
Comment (xylar) [reply uxarray/_mpas.py:208]:
```suggestion
    one_idx = (verticesOnCell.min() == 1)
```
This is currently working for completely the wrong reasons.  It is a one-item list, whereas it is intended to be a boolean.  Then, `if one_idx` evaluates to True later on because the list is not empty.

**Review by xylar — CHANGES_REQUESTED** [review]
I don't think the conversion to zero-based indexing is working for the right reasons, see inline comments below.

File: uxarray/_mpas.py
Comment (xylar) [inline uxarray/_mpas.py:89]: I just noticed these kinds of checks.  These also aren't going to work for the same reason I pointed out -- zeros as invalid values.

File: uxarray/_mpas.py
Comment (xylar) [inline uxarray/_mpas.py:158]: For the dual mesh, we need to talk about how to handle vertices with fewer than 3 neighboring cells.  These are indicated by zeros in `cellsOnVertex`.  The occur in planar or spherical meshes where some cells have been removed (e.g. land cells removed from the ocean).  For visualization purposes, a vertex with one or more invalid neighboring cells is degenerate.  But for mathematical purposes, they remain useful.  I don't know how uxarray plans to handle such situations.

In any case, some code needs to be added to convert zero values to mask values here.

**Comment by philipc2** [comment]
> how about point this to prototype branch rather than main? I think it will break a few things.
> 
> Good work with the changes for dual and prime mesh. We should document this in a notebook example.

We discussed the idea of linking this PR to the prototype branch during our Raijin meeting. We agreed that it would be best to keep this linked to the main branch, since the implementation and algorithms can be done disjoint from any of the redesign work. 

I think having a notebook example would be a great idea.

**Review by xylar — APPROVED** [review]
Excellent!  I think this looks right to me.  Nice testing!

File: uxarray/_mpas.py
Comment (philipc2) [reply uxarray/_mpas.py]: This is a great point. I think this would be a good candidate for a helper function.

Comment (xylar) [reply uxarray/_mpas.py]: I think your call to `_replace_zeros()` has taken care of this nicely.

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:229]:
```suggestion
    Returns
    -------
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:258]:
```suggestion
    Returns
    -------
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:254]:
```suggestion
    grid_var : numpy.ndarray
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:259]:
```suggestion
    grid_var : numpy.ndarray
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:278]:
```suggestion
    grid_var : numpy.ndarray
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:282]:
```suggestion
    Returns
    -------
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:283]:
```suggestion
    grid_var : numpy.ndarray
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:304]:
```suggestion
    use_dual : bool, optional
        Flag to select whether to encode the Dual-Mesh. Defaults to False
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:336]: Could we convert this print statement to an explicit error or warning?

File: uxarray/grid.py
Comment (anissa111) [inline uxarray/grid.py:214]: Why are you using `is not None` here? Wouldn't this theoretically produce unexpected results if `self.use_dual` is set, but set to False?

File: test/test_mpas.py
Comment (anissa111) [inline test/test_mpas.py:27]: Aren't these tests up here duplicating some work done by tests below? Do we need separate tests for execution and correctness since correctness will also fail if its execution fails?

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:226]:
```suggestion
    verticesOnCell : numpy.ndarray
        Vertex indices that surround a given cell

    nEdgesOnCell : numpy.ndarray
        Number of edges on a given cell
```

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:230]:
```suggestion
    verticesOnCell : numpy.ndarray
```

**Review by anissa111 — CHANGES_REQUESTED** [review]
Here's my first pass, thanks for working on this!

I still want to take a look at whether we can replace some of the for loops and numpy functions with more dask friendly alternatives, but I don't want to leave you waiting for a review for that

File: uxarray/_mpas.py
Comment (philipc2) [reply uxarray/_mpas.py:121]: Good idea, a warning would be appropriate.

File: uxarray/grid.py
Comment (philipc2) [reply uxarray/grid.py:214]: Since `use-dual` is part of the `**kwargs` of our `Grid` class, it's set to `None` if the user doesn't pass it through. See the usage example in the description for how it's supposed to be used.

File: uxarray/_mpas.py
Comment (philipc2) [reply uxarray/_mpas.py:125]: Good catch, thanks!

File: uxarray/grid.py
Comment (anissa111) [reply uxarray/grid.py:214]: Ah, yeah I see that now!

File: uxarray/_mpas.py
Comment (anissa111) [inline uxarray/_mpas.py:354]: I'm a bit concerned that this exception is broad enough, and there's enough that could go wrong in the try block, that this error message might not appropriately reflect what the root cause of an issue may be. 

I ran into this situation when I was running the tests while messing with some of the suggestions later in this review. I had made a typo in a variable name and the tests all triggered this exception message.

Would it be possible to wrap only the parts of `_dual_to_ugrid` and `_primal_to_ugrid` that are extracting the MPAS attributes in a try/catch instead of the entire read-in process?

**Review by anissa111 — COMMENTED** [review]
I'll approve after these are addressed!

File: uxarray/_mpas.py
Comment (xylar) [reply uxarray/_mpas.py:354]: Oh, I missed this but I agree with @anissa111 that the `try`/`except` is covering too much code.  In addition, it is best to avoid raising a generic `Exception` because as @anissa111 says, it catches *everything* including typos in the code or even someone trying to abort the code with `^c` (a `KeyboardInterrupt`).  If existing exceptions like `ValueError` or `OSError` aren't sufficient, you should strongly consider implementing your own exception.

Comment (philipc2) [reply uxarray/_mpas.py:354]: Good points, I didn't give this part enough attention (based it off what was already implemented in `_scrip.py`).

I'll work on setting up some more robust catches.

Comment (xylar) [reply uxarray/_mpas.py:354]: It might be better not to have this `try`/`except` at all and to focus on individual errors within each function that are more specific to the problems actually triggering the error.  As a user of `uxarray`, I wouldn't find this blanket exception very useful because it would replace a different exception that could point me toward the problem in more detail.

Comment (philipc2) [reply uxarray/_mpas.py:354]: I've removed the `try` / `except` entirely. 

Once we have more robust grid format checks, we should be able to handle errors and exceptions outside of these I/O modules. (see #238)

Example:
* User inputs a subset of an MPAS grid that excludes necessary variables to convert to UGRID
* We should be able to catch this before jumping into `_read_mpas()` and raise an appropriate exception

Comment (xylar) [reply uxarray/_mpas.py:354]: That sounds like a much cleaner plan  Thanks!

**Review by anissa111 — APPROVED** [review]

**Comment by philipc2** [comment]
@xylar 

Apologies for the re-request, you can ignore it (the grey checkmark confused me 😄) 

We need one more approval from anyone part of the Uxarray Group (@erogluorhan, @rajeeja , @rljacob) to get this merged.

**Review by xylar — APPROVED** [review]

**Comment by xylar** [comment]
No worries.  Happy to give you an approval that doesn't quite count :-)

Nice work on this!  No doubt there will be a few more hiccups to iron out as we try this in our workflows but it looks great to me.

**Review by rajeeja — APPROVED** [review]

**Pull request merged**
